packages/grid/_modules_/grid/components/panel/filterPanel/GridFilterInputSingleSelect.tsx

```diff
@@ -0,0 +1,111 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import TextField, { TextFieldProps } from '@mui/material/TextField';
+import { unstable_useId as useId } from '@mui/material/utils';
+import { GridLoadIcon } from '../../icons/index';
+import { GridFilterInputValueProps } from './GridFilterInputValueProps';
+import { GridColDef } from '../../../models/colDef/gridColDef';
+
+const renderSingleSelectOptions = ({ valueOptions, valueFormatter, field }: GridColDef, api) => {
+  const iterableColumnValues =
+    typeof valueOptions === 'function'
+      ? ['', ...valueOptions({ field })]
+      : ['', ...(valueOptions || [])];
+
+  return iterableColumnValues.map((option) =>
+    typeof option === 'object' ? (
+      <option key={option.value} value={option.value}>
+        {option.label}
+      </option>
+    ) : (
+      <option key={option} value={option}>
+        {valueFormatter && option !== '' ? valueFormatter({ value: option, field, api }) : option}
+      </option>
+    ),
+  );
+};
+
+export interface GridFilterInputSingleSelectProps extends GridFilterInputValueProps {
+  type?: 'singleSelect';
+}
+
+function GridFilterInputSingleSelect(props: GridFilterInputSingleSelectProps & TextFieldProps) {
+  const { item, applyValue, type, apiRef, focusElementRef, ...others } = props;
+  const [filterValueState, setFilterValueState] = React.useState(item.value ?? '');
+  const [applying, setIsApplying] = React.useState(false);
+  const id = useId();
+
+  const onFilterChange = React.useCallback(
+    (event) => {
+      let value = event.target.value;
+      // NativeSelect casts the value to a string.
+      const column = apiRef.current.getColumn(item.columnField);
+      const columnValueOptions =
+        typeof column.valueOptions === 'function'
+          ? column.valueOptions({ field: column.field })
+          : column.valueOptions;
+      value = columnValueOptions
+        .map((option) => (typeof option === 'object' ? option.value : option))
+        .find((optionValue) => String(optionValue) === value);
+
+      setFilterValueState(String(value));
+
+      setIsApplying(true);
```

Comment:
What is the use of `setIsApplying` is we set it to `true` then to `false` synchronously ?

Reply:
None. In my mind, `applyValue` could be time-consuming, and then be asynchronous. It's not the case so Yes, the setValue is useless here

Reply:
If it would be async then there is an `await` missing here.
But no, `applyValue` is just running `upsertFilterItem`, which can be heavy but is always synchronous according to it signature :+1: